File: order-service/src/main/java/github/lms/lemuel/shipping/domain/ShippingAddress.java
package github.lms.lemuel.shipping.domain;
import github.lms.lemuel.shipping.domain.exception.ShipmentInvariantViolationException;

import java.util.Objects;

/**
 * 배송지 값 객체. 한국 주소 체계 기준.
 */
public record ShippingAddress(
        String recipientName,
        String phone,
        String postalCode,
        String address1,
        String address2,
        String deliveryMemo
) {
    public ShippingAddress {
        Objects.requireNonNull(recipientName, "recipientName");
        Objects.requireNonNull(phone, "phone");
        Objects.requireNonNull(postalCode, "postalCode");
        Objects.requireNonNull(address1, "address1");
        if (recipientName.isBlank()) throw new ShipmentInvariantViolationException("수령인 이름 필수");
        if (phone.isBlank()) throw new ShipmentInvariantViolationException("전화번호 필수");
        if (postalCode.isBlank()) throw new ShipmentInvariantViolationException("우편번호 필수");
        if (address1.isBlank()) throw new ShipmentInvariantViolationException("주소 필수");
    }
}

File: frontend/src/lib/shippingAddress.ts
import { ShippingAddressRequest } from '@/types';

/**
 * 서버(`ShippingAddressSnapshot`)가 요구하는 필수 4항목. 비거나 공백뿐이면 400 이라
 * 화면에서 먼저 막는다.
 */
export const isShippingAddressComplete = (a: ShippingAddressRequest): boolean =>
  a.recipientName.trim() !== '' &&
  a.phone.trim() !== '' &&
  a.postalCode.trim() !== '' &&
  a.address1.trim() !== '';

export const emptyShippingAddress = (): ShippingAddressRequest => ({
  recipientName: '',
  phone: '',
  postalCode: '',
  address1: '',
  address2: '',
  deliveryMemo: '',
});
